# 📌 Funções e Métodos em Python para Ciência de Dados

  
Neste notebook, exploramos funções, métodos e técnicas de manipulação de dados em Python, com foco em aplicações práticas para Ciência de Dados.  
O conteúdo é aplicável para análise exploratória, limpeza de dados e preparação de datasets para modelagem.

**Objetivos:**
1. Criar funções reutilizáveis para análise de dados
2. Aplicar métodos nativos de Python para limpeza e transformação de dados
3. Usar funções `lambda` em Pandas para manipulação de colunas
4. Demonstrar boas práticas em Python voltadas para Data Science

## 1. Funções 
Exemplo: função para análise de métricas de vendas

In [5]:

def calcular_estatisticas(vendas):
    """
    Calcula métricas básicas de vendas
    
    Args:
        vendas (list): Lista de valores de venda
        
    Returns:
        dict: Dicionário com métricas
    """
    return {
        "total": sum(vendas),
        "media": sum(vendas)/len(vendas),
        "max": max(vendas),
        "min": min(vendas)
    }

# Testando
vendas_diarias = [150, 200, 80, 300]
calcular_estatisticas(vendas_diarias)

{'total': 730, 'media': 182.5, 'max': 300, 'min': 80}

## 2. Métodos em Python  
Exemplo de métodos úteis para Ciência de Dados:  
### Métodos de String

In [9]:
texto = "  Python Data Science  "
print(texto.strip())       # Remove espaços
print(texto.lower())       # Converte para minúsculas
print(texto.replace("Python", "R"))  # Substitui palavras
print(texto.split())       # Divide em lista de palavras


Python Data Science
  python data science  
  R Data Science  
['Python', 'Data', 'Science']


### Métodos de Lista

In [10]:
valores = [10, 5, 8]
valores.append(12)     # Adiciona elemento
valores.sort()         # Ordena a lista
valores.pop()          # Remove último elemento
valores


[5, 8, 10]

### Métodos de Dicionário

In [11]:
dados = {"produto": "Mouse", "preco": 89.90}
print(dados.keys())     # Chaves
print(dados.values())   # Valores
print(dados.items())    # Tuplas de chave-valor
print(dados.get("preco"))  # Acessa valor de uma chave


dict_keys(['produto', 'preco'])
dict_values(['Mouse', 89.9])
dict_items([('produto', 'Mouse'), ('preco', 89.9)])
89.9


## 3. Métodos para Limpeza de Dados  
#### Exemplo de limpeza e conversão de valores financeiros:

In [12]:
texto_bruto = "  RELATÓRIO: Vendas R$ 1.234,56  "

valor_limpo = (
    texto_bruto
    .strip()               # Remove espaços
    .split("R$")[-1]       # Pega texto após "R$"
    .strip()
    .replace(".", "")      # Remove pontos
    .replace(",", ".")     # Substitui vírgula
)

valor_numerico = float(valor_limpo)
valor_numerico


1234.56

##### Saída esperada:
1234.56

## 4. Funções Lambda em Python Puro  
 Exemplos de uso de lambda sem bibliotecas externas:  
#### Lambda em Lista

In [13]:
# Lista de preços
precos = [4250.90, 89.90, 199.90, 799.90]

# Adicionar frete de 10%
precos_com_frete = list(map(lambda x: x*1.10, precos))
precos_com_frete


[4675.99, 98.89000000000001, 219.89000000000001, 879.8900000000001]

#### Lambda com Condicional

In [14]:
# Aplicar desconto se preço > 100
precos_com_desconto = list(map(lambda x: x*0.9 if x>100 else x, precos_com_frete))
precos_com_desconto


[4208.391, 98.89000000000001, 197.901, 791.9010000000001]

#### Lambda com Dicionários

In [15]:
produtos = [
    {"nome": "Notebook", "preco": 4250.90},
    {"nome": "Mouse", "preco": 89.90},
    {"nome": "Teclado", "preco": 199.90},
]

# Atualizar preços aplicando frete
produtos_atualizados = list(map(lambda p: {**p, "preco_final": p["preco"] + 50}, produtos))
produtos_atualizados


[{'nome': 'Notebook', 'preco': 4250.9, 'preco_final': 4300.9},
 {'nome': 'Mouse', 'preco': 89.9, 'preco_final': 139.9},
 {'nome': 'Teclado', 'preco': 199.9, 'preco_final': 249.9}]